freezed_provider_value_notifier/lib/todo_list_model.dart

```diff
@@ -0,0 +1,131 @@
+// Copyright 2018 The Flutter Architecture Sample Authors. All rights reserved.
+// Use of this source code is governed by the MIT license that can be found
+// in the LICENSE file.
+
+import 'package:flutter/widgets.dart';
+import 'package:flutter/foundation.dart';
+import 'package:collection/collection.dart';
+import 'package:freezed_annotation/freezed_annotation.dart';
+import 'package:freezed_provider_value_notifier/models.dart';
+import 'package:todos_repository_core/todos_repository_core.dart';
+
+part 'todo_list_model.freezed.dart';
+
+enum VisibilityFilter { all, active, completed }
+
+@freezed
+abstract class TodoList with _$TodoList {
+  factory TodoList(
+    List<Todo> todos, {
+    @required VisibilityFilter filter,
+    @required bool loading,
+  }) = TodoListState;
+
+  @late
+  int get numCompleted =>
+      todos.where((Todo todo) => todo.complete).toList().length;
+
+  @late
+  bool get hasCompleted => numCompleted > 0;
+
+  @late
+  int get numActive =>
+      todos.where((Todo todo) => !todo.complete).toList().length;
+
+  @late
+  bool get hasActiveTodos => numActive > 0;
+
+  @late
+  List<Todo> get filteredTodos => todos.where((todo) {
+        switch (filter) {
+          case VisibilityFilter.active:
+            return !todo.complete;
+          case VisibilityFilter.completed:
+            return todo.complete;
+          case VisibilityFilter.all:
+          default:
+            return true;
+        }
+      }).toList();
+}
+
+extension TodoById on TodoList {
+  Todo todoById(String id) =>
+      todos.firstWhere((it) => it.id == id, orElse: () => null);
+}
+
+class TodoListController extends ValueNotifier<TodoList> {
+  TodoListController({
+    VisibilityFilter filter = VisibilityFilter.all,
+    @required this.todosRepository,
+    List<Todo> todos = const [],
+  })  : assert(todosRepository != null),
+        super(TodoList(todos, filter: filter, loading: false)) {
+    _loadTodos();
```

Comment:
General Q: Does it make sense to perform side effects in constructors? Overall, I feel like that's an "anti-pattern" since it often makes things like testing harder and means the user of the class has no control over when the fetch happens -- it just happens any time you construct this object.

IMO, it's better to make `loadTodos` public and call it from the `create` callback of a Provider.

Reply:
I don't see any issues personally.
The side-effect is controlled because we can mock `todosRepository` and I don't think the UI should be trigger side-effects once we've split things this way.

Reply:
Say you want to write this test:

```dart
test('should load todos from the repository', () async {
      final todos = [Todo('D')];
      final repository = MockRepository(todos);
      final model = TodoListController(todosRepository: repository);

      expect(
        model.value.todos,
        todos,
      );
    });
```

That's actually impossible with the current implementation, since you cannot await the load operation to complete.

The only way to make it work is to make `loadTodos` public and await it from the test function:

```dart
test('should load todos from the repository', () async {
      final todos = [Todo('D')];
      final repository = MockRepository(todos);
      final model = TodoListController(todosRepository: repository);

      await model.loadTodos();

      expect(model.value.todos, todos);
    });
```

Furthermore, the UI is actually already responsible for loading this data, but it's just happening implicitly by calling a constructor -- I'd argue that's really "surprising". I feel pretty strongly about this one: Constructors should construct an object, and that's it. They shouldn't be invoking a bunch of side effects. It's a recipe for confusion that I've dealt with quite a bit :/

Reply:
> That's actually impossible with the current implementation, since you cannot await the load operation to complete.

You could do:

```
await Future<void>.doWhile(() => Future.value(model.value.isLoading));
```

Reply:
Also, arguable with a `MockRepository` then the fetching is synchronous and a single event loop would be enough to skip it.

Also, `TodoListController` could expose a `Future<void> get done` (which does the previous `doWhile`)

Reply:
Furthermore, this is basically the same thing as what `change_notifier_provider` does, but instead of the constructor, it's moved to `addListener`.

Reply:
Oh my, you're actually right -- I forgot to change that part up from the original PR. IMO, that should definitely be removed and the provider code should look like this:

```dart
Provider<TodoListModel>(
      create: (_) => TodoListModel(repository: repository)..loadTodos(),
```

That said, I kinda feel like your argument is clever but doesn't address the overall point. Sure you could add a crazy `done` Future that performs a while loop until the state changes, but isn't that just exposing why performing this work in the constructor creates other issues? 

Dunno,  just seems a lot simpler to await the actual "loading done" Future directly.

Reply:
`TodoListModel` could also expose the raw future too.

It may even be easy to do with Dart 2.8:

```dart
class TodoListModel {
  late final todos = _loadTodos();

  Future<List<Todo>> _loadTodos() => ...
}
```

Then tests could just `await model.todos`.

Overall, I would disagree that this can cause problems

Reply:
No worries -- agree to disagree.

I'd only ask that you add a test like the one above to the test suite for the controller, since interacting with the repo is a big part of it's job  (was missing that test in the change notifier provider sample, added one here: https://github.com/brianegan/flutter_architecture_samples/pull/176/files#diff-26b067bb8f7860e5e48b6b1edd3bdd49).

Reply:
Done